Add option to directly open fullscreen when the main player starts

Stypox opened the pull request:
<!-- Hey there. Thank you so much for improving NewPipe, and filling out the details. Having roughly the same layout helps everyone considerably :)-->

#### What is it?
- [ ] Bugfix (user facing)
- [x] Feature (user facing)
- [ ] Codebase improvement (dev facing)
- [ ] Meta improvement to the project (dev facing)

#### Description of the changes in your PR
This PR adds a switch to video&audio settings that, when enabled, causes the main player to start automatically in fullscreen mode. This was rather painful to implement, but I think this finally works well.

#### Fixes the following issue(s)
Fixes #4152
Fixes #4387
Fixes #4682
Fixes #5845
Fixes #5594
Fixes #5299
Fixes #6785
Fixes #6784
Fixes #6131
See #4548

#### APK testing 
The APK can be found by going to the "Checks" tab below the title. On the left pane, click on "CI", scroll down to "artifacts" and click "app" to download the zip file which contains the debug APK of this PR.

#### Due diligence
- [x] I read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md).

opusforlife2 commented on app/src/main/res/values/strings.xml:
```diff
@@ -571,9 +571,10 @@
+    <string name="autoplay_and_fullscreen_summary">Start playback automatically — %s</string>
     <string name="wifi_only">Only on Wi-Fi</string>
     <string name="never">Never</string>
+    <string name="never_and_start_in_fullscreen">Never, and directly enter fullscreen when tapping on the thumbnail if applicable</string>
```
```suggestion
    <string name="never_and_start_in_fullscreen">Never, and directly enter fullscreen when tapping on the thumbnail</string>
```

opusforlife2 commented:
Well, first try: the rotation is quite laggy and jerky. The player first opens in portrait, the video starts loading, and then after some time, the player rotates. This can cause the loading circle to go to a corner until the video appears.

Can the player first rotate, and only after that's completed the video starts loading?

Stypox commented:
@opusforlife2 I can try, though the player code is so complex that I don't know if I will make it

nikhilCad commented:
Any updates?

Stypox commented:
@avently sorry to bother you. Do you have any solution for this? That is, opening the fullscreen player directly?

> Well, first try: the rotation is quite laggy and jerky. The player first opens in portrait, the video starts loading, and then after some time, the player rotates. This can cause the loading circle to go to a corner until the video appears.
>
> Can the player first rotate, and only after that's completed the video starts loading?

avently commented:
>sorry to bother you

no problem.


Add
```
if (shouldPlayInFullscreen) {
onScreenRotationButtonClicked()
}
```
here before openVideoPlayer:
https://github.com/TeamNewPipe/NewPipe/blob/dev/app/src/main/java/org/schabi/newpipe/fragments/detail/VideoDetailFragment.java#L495

and in other similar methods. there is no need to do any huge code change. only those you made for additional autoplay option.
 `shouldPlayInFullscreen` is a sharedPreference option that is your new autoplay option.

does it work like you expect?

Stypox commented:
I rebased and implemented what @avently suggested (though some testing is still needed). I also fixed the review by @opusforlife2

nikhilCad commented:
@avently friendly pin to check it now that @Stypox had made some changes

avently commented:
@Stypox
> TODO does it work for large-land layout? 

Nope, it doesn't. Because there is a check that prevent landscape layouts from advantage to go to fullscreen. So probably you need to specify
```
(!isLandcspe || isTablet)
```
To test the changes on large-land you could specify in developer settings of the phone/emulator Device width like 600

avently commented:
And maybe to call fullscreenButtonClicked in this case, not sure

Stypox commented:
@avently could you please implement it? I wasn't able to reach a solution, I've been trying for two hours and I ran into multiple inconsistent and strange behaviours, but I didn't find a solution working for every situation.

avently commented:
Ok, i'll look

opusforlife2 commented:
> I rebased and implemented what avently suggested (though some testing is still needed). The fluidity is still not perfect, but it wasn't so even in pre-unified, so I don't think it can be improved much. I also fixed the review by opusforlife2

@Stypox MUCH better than before, thank you. It's not perfect, like you said, but at least the video only starts after the app goes to full screen, instead of showing a black screen and a distorted player layout, forcing you to rewind the video from the beginning.

For reference, the current behaviour is: Tap thumbnail -> App goes into landscape (still on video details) -> The video stream is being fetched -> The moment it starts playing, the player goes into full screen.

Intended perfect behaviour (IMO): Tap thumbnail -> App goes into landscape AND player goes into full screen -> The video stream is being fetched -> The video starts playing. (Similar to 0.19.x behaviour).

opusforlife2 commented:
@Stypox Incidentally, the ripple animation when tapping on the thumbnail is gone, leaving no visual feedback of the tapping action. Intentional?

Stypox commented:
> Ok, i'll look

Thank you, and sorry again for bothering you ;-)

> Incidentally, the ripple animation when tapping on the thumbnail is gone, leaving no visual feedback of the tapping action. Intentional?

No, that's probably just another random bug caused by the messiness of the code

avently commented:
@opusforlife2 

Remove
```
&& !player.isPlaying()
```
Here
https://github.com/TeamNewPipe/NewPipe/blob/fd68a1e57c9c510c673afc15f450d940d0e2dc15/app/src/main/java/org/schabi/newpipe/fragments/detail/VideoDetailFragment.java#L2030

Ane look at the result

And the app will be in fullscreen right after orientation change. Instead of removing this immediate fullscreen can be turned on for those who have autofullscreen enabled instead of everyone

> No, that's probably just another random bug caused by the messiness of the code

No, that's because you're forcing activity to change the orientation. It do it in main thread, ripple only works when main thread is free. So nothing strange here, immediate feedback of ui

avently commented:
@Stypox I played with the code and looks like I found the best possible solution. However I didn't tested all million possible ways to interact with the app. But I like what I see. You probably want to move your code from onClickListener to the new place when you'll see how awesome it works on tablets.

Ok, add this lines
```
if (PlayerHelper.getAutoplayType(service)
                == PlayerHelper.AutoplayType.AUTOPLAY_TYPE_NEVER_AND_START_IN_FULLSCREEN
                && DeviceUtils.isTablet(service)
                && fragmentListener != null
                && videoPlayerSelected()
                && PlayerHelper.globalScreenOrientationLocked(service)) {
            fragmentListener.onScreenRotationButtonClicked();
        }
```
Here
 https://github.com/TeamNewPipe/NewPipe/blob/fd68a1e57c9c510c673afc15f450d940d0e2dc15/app/src/main/java/org/schabi/newpipe/player/Player.java#L600

Better to have a separate method of course and just to call it here.
Good?

Stypox commented:
@avently thank you for your dedication! I pushed an update, could you check if it is fine code-wise? I tested on phone, tablet and TV and everything seems fine.

avently commented:
After moving the code from onClick to openVideoPlayer the flow is broken. See a screen record. Also this change has undesireable behaviour when pressing on play button in mini player (but it's questionable, maybe it's ok).

https://user-images.githubusercontent.com/7953703/112749383-bb17d880-8fca-11eb-9f42-315c1e3d81ab.mp4

Stypox commented:
@avently I reverted that change and tested a little bit, everything looks fine. Vertical videos turn around to landscape even though they look best in portrait, but that's expected as this feature is mostly for TVs (or tablets) where turning around the device is impossible (or cumbersome), and anyway that's the expected behaviour.

peat80 commented:
Just testing latest build on my shieldtv atm. with a few videos. Seems to work fine. Good job. 😀👍

peat80 commented:
Works really fine and hopefully will make it into one of the next releases. 👍

Would it be possible to add another option?

(Yes, and directly enter fullscreen)

Mostly if you click on a video in your feed on the tv one would directly start playing it in fullscreen on the tv.
The less clicks the better. 😀👍

Stypox commented:
@peat80 then I think we should just revert adding "Never, and directly..." and instead add a "Start main player in fullscreen" option. What do you think @TiA4f8R @opusforlife2 ?

avently commented:
And how it will work then in details?

Stypox commented:
@avently with that option enabled it will start directly in fullscreen if:
- the video detail fragment has just been opened (i.e. a stream has been clicked) and autoplay kicks in (i.e. either it is set to Always or it is set to WiFi and wifi is available)
- the user taps on the thumbnail

peat80 commented:
> @avently with that option enabled it will start directly in fullscreen if:
> 
> * the video detail fragment has just been opened (i.e. a stream has been clicked) and autoplay kicks in (i.e. either it is set to Always or it is set to WiFi and wifi is available)
> * the user taps on the thumbnail

That would be the best solution probably. 👌

opusforlife2 commented:
Well, firstly a bug: the latest APK pauses videos as soon as they start. This tends to be useful in certain cases, but a problem in most others.

---
Regarding the auto full screen behaviour: is there no way to let the player detect the video orientation and _then_ go into full screen, rotating if needed? Or is that too late in the process?

The current behaviour of the PR, though not optimal, is at least _equivalent_ to the one present in 0.19.x. Even back then, if you came across a portrait video among all the landscape ones, you would have to toggle the orientation at least once. The same thing happens here. You have to tap the full screen button once to go into full screen portrait mode.

Some odd behaviour (which actually ends up helping instead of hurting): Tapping the full screen button while a portrait video is being played in landscape corrects the orientation of full screen to portrait, without going back to the embedded player in the middle. From that point, tapping the button keeps switching between portrait full screen and portrait embedded player.

So from my point of view, at this stage this PR is at least _functionally_ restoring the 0.19.x behaviour, which is great.

If there could be a way to automatically go into full screen, but in the correct orientation without any user interaction, that would be out of this world.

futureisfoss commented:
> @avently: I'm curious where all these people who screams about auto fullscreen here and there? So when somebody really doing something that they're asking of they just keep silence. It gives fully understanding that such people is here for screaming (not for real profit of the app). This is what I was confident in since the beginning. If you give power to them (I mean if you do something you don't like just because they ask) it would make the app worse only. Just a sidenote.

I'm quite new on github, so I wasn't one of those people who screamed about "auto fullscreen". But I was one of the fellow users who were really sad with the loss of this feature. It may look like a few more extra clicks, but when you multiply it with the number of videos you watch, it becomes very inefficient.

So, I'm just reassuring you: This actually is a really good feature and is totally worth the effort. Thanks to everyone who contributed to this ❤️

SameenAhnaf commented:
Would you guys please consider implementing https://github.com/TeamNewPipe/NewPipe/issues/5675 ? That will be highly appreciated. Many users like me only need full screen video.

SameenAhnaf commented:
I am not sure if this is the correct place to ask. But there should be an option for continuing to main player even if 'Auto-play' option 'Never' is selected.

 What do you guys think about this solution?
Click on 'Play' icon on thumbnail: Switch to full screen
Click elsewhere on the thumbnail: Start playing on main player

![IMG_20210502_002133](https://user-images.githubusercontent.com/79245641/116791462-617c5f80-aadc-11eb-9fd0-a3517b8a738b.jpg)

andrewferguson commented:
I know this will probably be seen as "yet another user complaining about *that* problem", but I wanted to explain briefly why the auto-full-screen functionality as it was in 0.19.x was so valuable, at least in my case.

My phone is cheap and..... not particularly great. In particular, the touch screen isn't very accurate. Tapping on the video thumbnail to begin full-screen playback was great - a simple one-tap operation. Tapping on the thumbnail *and then* needing to tap on the full-screen icon is much, much more difficult. I often have to tap two or three times before it has registered, and by the time it enters full screen I have missed the first 5 or so seconds of the video.

It's coming up on 9 months since version 0.20 was released, and I still haven't "adapted" to this new workflow - every time I have to play a video, the lack of an ability to tap the thumbnail and have the video start playing in full screen feels really clumsy to me. I completely agree with the points raised about avoiding disrupting new users of the 0.20.x release by making this a default-off feature, but I also think that by excluding the feature entirely the UX is reduced for the (possibly small) subset of users who would appreciate this feature.

Having said all the above, I do realise the pressures that the NewPipe team are under. This is a hugely popular project, and you are doing a commendable job putting out regular releases. The last thing I want to do is burden you with extra work for an issue that probably only affects a small (but vocal!) number of users. Reading through this pull request has made me realise how much more complex such features are - supporting tablets and Android TV was something that never even occurred to me!

I do know a bit of Java, but have never done Android programming. I suppose the first question to try and get this resolved would be to ask @opusforlife2 and @Stypox about how complete this PR was when work seemed to stop in April. With all the new features added since then, is merging this PR still feasible? (I see the GitHub UI is warning of quite a few conflicts).

peat80 commented:
Would be indeed awesome if this pr could be finished and then merged in the near future. Waiting for this since ages. 😃👌

opusforlife2 commented:
@andrewferguson Conflicts are almost inevitable when a PR is left unattended for a while. It's normal. As for feasibility, Stypox is usually the one who takes on _other people's_ PRs and finishes the work for merging. If Stypox is busy... then you'll probably have to wait, as will I. :P

andrewferguson commented:
This is fantastic! Thanks so much @Stypox for getting this change ready,

> I tested on my phone API 29, on emulator API 19 and on emulator API 30 and everything seemed to work fine. Could you test, too? @opusforlife2 @peat80 @andrewferguson

Sure, I'll give it a try. Time for a n00b question: I assume I'll need to setup the Android development tools and build the APK from the master branch?

TiA4f8R commented:
@andrewferguson No, you don't have to build yourself the code: use the check `build-and-test-jvm` from the `Checks` tab (of the tabs at the top of this PR) to get a debug APK generated by GitHub actions.

peat80 commented:
Works exactly as it should. Great job. 😃👌

Only thing i noticed is when you exit to the launcher (so that it switches to background or popup play), then go back into newpipe starting a video it does not open in fullscreen and plays in normal tablet mode.

luminous706 commented:
Great! I was looking for this feature, I'm glad this is being worked on! 

My use-case: I share links to a secondary phone that my toddler holds, I select what he watches and these videos should start fullscreen.

Stypox commented:
> Only thing i noticed is when you exit to the launcher (so that it switches to background or popup play), then go back into newpipe starting a video it does not open in fullscreen and plays in normal tablet mode.

I can't reproduce, could you provide reproducing steps and possibly a video?

peat80 commented:
@ #

https://user-images.githubusercontent.com/32403292/127198707-6dc497f4-1494-4514-8142-e447a2aed10a.mp4

Start newpipe > start video > exit to homescreen with home key (circle on shieldtv) > configured background playback starts >switch back into newpipe (miniplayer at bottom shows background still playing) navigate to any video again > it opens in tablet mode.

andrewferguson commented:
It's really great to see this feature back in NewPipe (albeit in a beta / testing version). A few comments (note that all of these comments are from the perspective of someone who uses their phone with the orientation set to "portrait"):

1. A minor thing, but the UX of having the screen go black, then the video info screen opens in landscape mode, then the video start playing in full screen, is a little clumsy. It would be nice if the second step (video info screen opens in landscape mode) was eliminated, meaning the screen goes black, then the video starts playing in full screen mode. Of course, this is a minor nuisance, and might be impossible to fix if the player library used by NewPipe can't support it, but I thought I'd mention it anyway.
2. When a video in full-screen mode stops playing because the end of the video is reached, the player exits full-screen mode but does not return to the portrait orientation. I need to manually press the "full screen" button in the player in order to return to the portrait orientation. However, if I press the Android "back" button before the video ends, the player does automatically return to the portrait orientation.
3. There is a bug on my device when I exit to the home screen whilst playing a video in full-screen mode. I get either the error "Unrecoverable player error occurred" or "Recovering from player error". Despite this, when I return to NewPipe from the home screen, there doesn't seem to be any 'unrecoverable' error - I can continue playing the video without issue. I have made a screen recording of this issue, in the video newpipe-test-4.mp4.
4. With the same setup as in point 3., when I return to a video from the home screen, my expectation is that the video should be in the full screen mode, since this is how I left the video. But the video is not in the full-screen mode, just the landscape orientation. This issue can also be seen in the video newpipe-test-4.mp4
5. There is a bug on my device when I navigate to the "app switcher" whilst playing a video in full-screen mode. When I return to NewPipe through the app switcher, the video is in a paused state, but when I attempt to play it only the audio begins playing and the video is frozen. I have made a screen recording of this issue, in the video newpipe-test-5.mp4. (There is no audio in the screen recording because the screen recording app does not support recording audio).
6. There is a bug on my device when I play a video that I have just played. I get an "Unrecoverable player error occured" message, although tapping the video allows playback to continue. However, if I then try to exit using the Android "Back" button, it won't let me exit. Repeatedly pressing the back button eventually allows me to exit from the full-screen player, however unlike in normal situations when the "Back" button returns in the portrait orientation, this exit from the player is still in the horizontal orientation. I have made a screen recording of this issue, in the video newpipe-test-6.mp4. 

To be fair, all of these are fairly minor issues that can be worked around. In particular, 2. is one that exists because that's the current behaviour of NewPipe - if you're watching a video in full-screen mode and it ends, you stay in landscape, even if tapping the full-screen button at the beginning of video playback was what took you into landscape orientation. None of these issues are showstoppers, and none dampen the excitement I had this morning when I read my emails and realised that this fix was significantly closer to making it into a release.

My thanks to @Stypox for working to make this happen!

https://user-images.githubusercontent.com/13520045/127218628-e19ae365-a99a-4feb-ac0a-545d35745027.mp4

https://user-images.githubusercontent.com/13520045/127218741-63a8e4f0-1899-4e30-8291-b6a7299417fd.mp4

https://user-images.githubusercontent.com/13520045/127218751-a5184917-9e27-4b96-88eb-cf8d8b0f1e8f.mp4

infinitewaveparticle commented:
> 2. When a video in full-screen mode stops playing because the end of the video is reached, the player exits full-screen mode but does not return to the portrait orientation. I need to manually press the "full screen" button in the player in order to return to the portrait orientation. However, if I press the Android "back" button before the video ends, the player does automatically return to the portrait orientation.

This is definitely something I've noticed for a while. I think for users who wish to browse in portrait orientation, but watch in fullscreen landscape orientation it would be best after the end of a video to either return to the portrait mode automatically or be able to hit back and go to portrait mode automatically. I usually have to restart the video and THEN hit back to get back to portrait mode. My phone is locked to portrait mode, btw.

Stypox commented:
Thank you for the effort of finding out all of those bugs. In order for this PR to be merged soon, I'd not address them here (they all appear in the current version of NewPipe, too, don't they? So they are not caused by this PR).

opusforlife2 commented:
Yay! Thanks, Stypox!

There is a lag of ~3-4 seconds between tapping the thumbnail and the rotation. Apart from that, there is no visual feedback upon tapping the thumbnail, which could lead the user to think something is wrong and start trying things.

> I could not reproduce with the current code.

Yeah that turned out to be a bug in all the debug APKs at the time. Don't remember when it was fixed.

Stypox commented:
> There is a lag of ~3-4 seconds between tapping the thumbnail and the rotation. Apart from that, there is no visual feedback upon tapping the thumbnail, which could lead the user to think something is wrong and start trying things.

There is nothing I can do about that unfortunately, or at least not in this PR :disappointed:

opusforlife2 commented:
Nyo problem. Another PR, then. 👍

SameenAhnaf commented:
#6131 should also be linked with the PR.

TobiGr commented on app/src/main/java/org/schabi/newpipe/fragments/detail/VideoDetailFragment.java:
```diff
@@ -1112,6 +1130,12 @@ public void openVideoPlayer() {
     }
 
+    public void openVideoPlayerAutoFullscreen() {
+        // if the option to start directly fullscreen is enabled, openVideoPlayer will be called
+        // with directlyFullscreenIfApplicable=true and therefore open fullscreen if applicable
```
Please use JavaDoc

TobiGr commented on app/src/main/java/org/schabi/newpipe/player/Player.java:
```diff
@@ -742,6 +745,20 @@ public void handleIntent(@NonNull final Intent intent) {
+                && PlayerHelper.isStartMainPlayerFullscreen(service)
+                && DeviceUtils.isTablet(service)
+                && videoPlayerSelected()
+                && PlayerHelper.globalScreenOrientationLocked(service)) {
```
Why do we only need the last condition? The player does not start in fullscreen mode when using the tablet emulator with auto-rotation and start main player in fullscreen enabled. That's irritating and unexpected behaviour for me.

avently replied:
Because you can rotate the device and get fullscreen easily with unlocked autorotation. Totally expected like everywhere in the app

Stypox replied:
Yes, this is the standard behaviour. I'll explain this is the case in the setting description.

opusforlife2 commented:
> I tested this PR now found unexpected error.

@avently I followed your steps and didn't get the bug. Can you try again? If you still see it, please mention what video player settings you're using.

avently commented:
Video demonstration of https://github.com/TeamNewPipe/NewPipe/pull/5459#issuecomment-907433431
Happens with default settings too. On the video I just changed language and content language just for your better understanding.
https://user-images.githubusercontent.com/7953703/131280656-a8599183-2843-470d-becf-298e3d9c78f0.mp4

Stypox commented:
```logcat
E/ExoPlayerImplInternal: Playback error
      com.google.android.exoplayer2.ExoPlaybackException: MediaCodecVideoRenderer error, index=0, format=Format(1, null, null, video/avc, null, -1, null, [1280, 720, 23.976023], [-1, -1]), format_supported=YES
        at com.google.android.exoplayer2.ExoPlayerImplInternal.handleMessage(ExoPlayerImplInternal.java:542)
        at android.os.Handler.dispatchMessage(Handler.java:103)
        at android.os.Looper.loop(Looper.java:214)
        at android.os.HandlerThread.run(HandlerThread.java:67)
     Caused by: com.google.android.exoplayer2.mediacodec.MediaCodecRenderer$DecoderInitializationException: Decoder init failed: OMX.qcom.video.decoder.avc, Format(1, null, null, video/avc, null, -1, null, [1280, 720, 23.976023], [-1, -1])
        at com.google.android.exoplayer2.mediacodec.MediaCodecRenderer.maybeInitCodecWithFallback(MediaCodecRenderer.java:1052)
        at com.google.android.exoplayer2.mediacodec.MediaCodecRenderer.maybeInitCodecOrBypass(MediaCodecRenderer.java:609)
        at com.google.android.exoplayer2.mediacodec.MediaCodecRenderer.onInputFormatChanged(MediaCodecRenderer.java:1480)
        at com.google.android.exoplayer2.video.MediaCodecVideoRenderer.onInputFormatChanged(MediaCodecVideoRenderer.java:640)
        at com.google.android.exoplayer2.mediacodec.MediaCodecRenderer.readToFlagsOnlyBuffer(MediaCodecRenderer.java:999)
        at com.google.android.exoplayer2.mediacodec.MediaCodecRenderer.render(MediaCodecRenderer.java:849)
        at com.google.android.exoplayer2.ExoPlayerImplInternal.doSomeWork(ExoPlayerImplInternal.java:892)
        at com.google.android.exoplayer2.ExoPlayerImplInternal.handleMessage(ExoPlayerImplInternal.java:467)
        at android.os.Handler.dispatchMessage(Handler.java:103) 
        at android.os.Looper.loop(Looper.java:214) 
        at android.os.HandlerThread.run(HandlerThread.java:67) 
     Caused by: java.lang.IllegalArgumentException: The surface has been released
        at android.media.MediaCodec.native_configure(Native Method)
        at android.media.MediaCodec.configure(MediaCodec.java:2026)
        at android.media.MediaCodec.configure(MediaCodec.java:1954)
        at com.google.android.exoplayer2.mediacodec.SynchronousMediaCodecAdapter.configure(SynchronousMediaCodecAdapter.java:43)
        at com.google.android.exoplayer2.video.MediaCodecVideoRenderer.configureCodec(MediaCodecVideoRenderer.java:580)
        at com.google.android.exoplayer2.mediacodec.MediaCodecRenderer.initCodec(MediaCodecRenderer.java:1148)
        at com.google.android.exoplayer2.mediacodec.MediaCodecRenderer.maybeInitCodecWithFallback(MediaCodecRenderer.java:1045)
        at com.google.android.exoplayer2.mediacodec.MediaCodecRenderer.maybeInitCodecOrBypass(MediaCodecRenderer.java:609) 
        at com.google.android.exoplayer2.mediacodec.MediaCodecRenderer.onInputFormatChanged(MediaCodecRenderer.java:1480) 
        at com.google.android.exoplayer2.video.MediaCodecVideoRenderer.onInputFormatChanged(MediaCodecVideoRenderer.java:640) 
        at com.google.android.exoplayer2.mediacodec.MediaCodecRenderer.readToFlagsOnlyBuffer(MediaCodecRenderer.java:999) 
        at com.google.android.exoplayer2.mediacodec.MediaCodecRenderer.render(MediaCodecRenderer.java:849) 
        at com.google.android.exoplayer2.ExoPlayerImplInternal.doSomeWork(ExoPlayerImplInternal.java:892) 
        at com.google.android.exoplayer2.ExoPlayerImplInternal.handleMessage(ExoPlayerImplInternal.java:467) 
        at android.os.Handler.dispatchMessage(Handler.java:103) 
        at android.os.Looper.loop(Looper.java:214) 
        at android.os.HandlerThread.run(HandlerThread.java:67) 
```

Stypox commented:
- @avently video isn't playing and I see thumbnail - Fixed in 45599c7190065357caab58948343b82996fef533 by completely closing the player when switching video detail with autoplay disabled
- @TobiGr Please use JavaDoc - Done in b8f16d0a76a858d0a02b0141b4de02ca93e459f5
- I'll explain this is the case in the setting description - Done
- Also fixed a NPE that happened while testing - ab97f485825c723036ee2d10c16a785e6b1e5d4f

TobiGr commented on app/src/main/java/org/schabi/newpipe/player/helper/PlayerHelper.java:
```diff
@@ -239,6 +239,11 @@ public static boolean isBrightnessGestureEnabled(@NonNull final Context context)
                 .getBoolean(context.getString(R.string.brightness_gesture_control_key), true);
     }
 
+    public static boolean isStartMainPlayerFullscreen(@NonNull final Context context) {
```
?

TobiGr commented on app/src/main/res/xml/video_audio_settings.xml:
```diff
@@ -129,6 +129,13 @@
+            android:key="@string/start_main_player_fullscreen_key"
+            android:summary="@string/start_main_player_fullscreen_summary"
+            android:title="@string/start_main_player_fullscreen_title"
+            app:iconSpaceReserved="false" />
```
```suggestion
            app:singleLineTitle="false"
            app:iconSpaceReserved="false" />
```

Stypox replied:
Mmmh, isn't there a way to set these attributes in a theme of some sort, so that they do not need to be specified each time?

Stypox commented:
- Renamed `isStartMainPlayerFullscreen` to `isStartMainPlayerFullscreenEnabled` (sorry, I forgot about this @TobiGr)
- Added `app:singleLineTitle="false"`

tsiflimagas commented:
There seems to be a problem with this. Tapping on the button to go to main player from popup, it switches to fullscreen, even when the option is disabled.

Stypox commented:
@tsiflimagas yes, that's expected. The symbol on the popup fullscreen button is a fullscreen one, hence it should behave as fullscreen.

tsiflimagas commented:
@Stypox ohh, that's unfortunate for me :) Now there's no way on the popup player to open normally in the main player, so the only way is to open the app and tap on the main player (which has the problem of moving the seekbar to where it was when the video was switched to popup player, not where it was a moment ago (idk if there's already an issue for this)). It felt wrong when the app was rotated, while I just wanted to switch player (and landscape mode is really pain in the neck on these long and narrow phones). I hope something new will come up for this.
